tests/Helper/RulesNormalizerTest.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Tests\Helper;
+
+use PHPUnit\Framework\TestCase;
+use ReflectionProperty;
+use Yiisoft\Validator\Helper\RulesNormalizer;
+use Yiisoft\Validator\Tests\Support\Data\ObjectWithDifferentPropertyVisibility;
+
+final class RulesNormalizerTest extends TestCase
+{
+    public function dataNormalize(): array
+    {
+        return [
+            'null' => [[], null],
+            'default-property-visibility' => [
+                [
+                    'name' => ['required'],
+                    'age' => ['number'],
+                    'number' => ['number'],
+                ],
+                new ObjectWithDifferentPropertyVisibility(),
+            ],
+            'private-properties' => [
+                [
+                    'number' => ['number'],
+                ],
+                new ObjectWithDifferentPropertyVisibility(),
+                ReflectionProperty::IS_PRIVATE,
+            ],
+            'protected-properties' => [
+                [
+                    'age' => ['number'],
+                ],
+                new ObjectWithDifferentPropertyVisibility(),
+                ReflectionProperty::IS_PROTECTED,
+            ],
+            'public-properties' => [
+                [
+                    'name' => ['required'],
+                ],
+                new ObjectWithDifferentPropertyVisibility(),
+                ReflectionProperty::IS_PUBLIC,
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider dataNormalize
+     */
+    public function testNormalizeWithArrayResult(
+        array $expected,
+        iterable|object|string|null $rules,
+        ?int $propertyVisibility = null,
+        mixed $data = null
+    ): void {
+        $rules = $propertyVisibility === null
```

Comment:
Contradicts with recent discussion from another PR - https://github.com/yiisoft/validator/pull/373#discussion_r1032923815. But I'm OK with this approach.

Reply:
:man_shrugging: OK.